File: app/src/main/java/pl/coddev/applu/c/Prefs.java
package pl.coddev.applu.c;

import android.content.Context;
import android.content.SharedPreferences;

import pl.coddev.applu.AppluApplication;
import pl.coddev.applu.enums.AppSelectorStatus;

public class Prefs {

    public static SharedPreferences prefs = AppluApplication.get()
            .getSharedPreferences(Constants.PREF_FILE, Context.MODE_PRIVATE);
    public static final String LAUNCHED_BEFORE = "launchedBefore";
    private static Prefs instance;

    private Integer featureCount;
    private String appSelectorStatus;
    private Integer appIndex;
    private String currentApp;
    private String filterList;
    private String lastApps;
    private String lastAppsSync;

    public static Prefs get() {
        if (instance == null) return new Prefs();
        else return instance;
    }

    public static SharedPreferences.Editor getEdit() {
        return prefs.edit();
    }

    public int getFeatureCount() {
        if (featureCount == null) {
            featureCount = prefs.getInt(Constants.EXTRA_FEATURE_USAGE, 0);
        }
        return featureCount;
    }

    public void setFeatureCount(int featureCount) {
        this.featureCount = featureCount;
        putInt(Constants.EXTRA_FEATURE_USAGE, featureCount);
    }

    public AppSelectorStatus getAppSelectorStatus(int widgetId) {
        if (appSelectorStatus == null) {
            appSelectorStatus = prefs.getString(Constants.APP_SELECTOR_STATUS + widgetId,
                    AppSelectorStatus.USER.name());
        }
        return AppSelectorStatus.valueOf(appSelectorStatus);
    }

    public void setAppSelectorStatus(AppSelectorStatus appSelectorStatus, int appWidgetId) {
        this.appSelectorStatus = appSelectorStatus.name();
        putString(Constants.EXTRA_FEATURE_USAGE + appWidgetId, appSelectorStatus.name());
    }

    public Integer getAppIndex(int widgetId) {
        if (appIndex == null) {
            appIndex = prefs.getInt(Constants.APP_INDEX + widgetId, 0);
        }
        return appIndex;
    }

    public void setAppIndex(Integer appIndex, int widgetId) {
        this.appIndex = appIndex;
        putInt(Constants.APP_INDEX + widgetId, appIndex);
    }

    public void setCurrentApp(String currentApp, int widgetId) {
        this.currentApp = currentApp;
        putString(Constants.EXTRA_FEATURE_USAGE + widgetId, currentApp);
    }

    public String getCurrentApp(int widgetId) {
        if (currentApp == null) {
            currentApp = prefs.getString(Constants.CURRENT_APP + widgetId, "");
        }
        return currentApp;
    }

    public String getFilterList(int widgetId) {
        if (filterList == null) {
            filterList = prefs.getString(Constants.PREFS_FILTER_LIST + widgetId, "");
        }
        return filterList;
    }

    public void setFilterList(String filterList, int widgetId) {
        this.filterList = filterList;
        putString(Constants.PREFS_FILTER_LIST + widgetId, filterList);
    }

    public String getLastApps(int widgetId) {
        if (lastApps == null) {
            lastApps = prefs.getString(Constants.LAST_APPS + widgetId, "");
        }
        return lastApps;
    }

    public void setLastApps(String lastApps, int widgetId) {
        this.lastApps = lastApps;
        putString(Constants.LAST_APPS + widgetId, lastApps);
    }

    public String getLastAppsSync(int widgetId) {
        if (lastAppsSync == null) {
            lastAppsSync = prefs.getString(Constants.LAST_APPS_SYNC + widgetId, "");
        }
        return lastAppsSync;
    }

    public void setLastAppsSync(String lastAppsSync, int widgetId) {
        this.lastAppsSync = lastAppsSync;
        putString(Constants.LAST_APPS_SYNC + widgetId, lastAppsSync);
    }

    public static void remove(String... keys) {
        for (String key : keys) {
            remove(key);
        }
    }

    public static String getString(String key) {
        return prefs.getString(key, null);
    }

    public static void remove(String key) {
        prefs.edit().remove(key)
                .apply();
    }

    public static void putBool(String key, boolean value) {
        prefs.edit().putBoolean(key, value).apply();
    }

    public static void putString(String key, String value) {
        prefs.edit().putString(key, value).apply();
    }

    public static void putInt(String key, int value) {
        prefs.edit().putInt(key, value).apply();
    }

    public static boolean bool(String key) {
        return prefs.getBoolean(key, false);
    }

    public static void clear() {
        prefs.edit().clear().apply();
    }

    public static boolean runningThisFirstTime(String tag) {
        if (!prefs.getBoolean(tag, false)) {
            prefs.edit().putBoolean(tag, true).apply();
            return true;
        } else {
            return false;
        }
    }

}

File: app/src/main/java/pl/coddev/applu/utils/Prefs.kt
package pl.coddev.applu.utils

import android.content.Context
import android.content.SharedPreferences
import pl.coddev.applu.App
import java.util.*

object Prefs {

    const val LAST_APPS_MAX_SIZE = 6
    const val LAUNCHED_BEFORE = "launchedBefore"
    private var prefs: SharedPreferences = App.get()
            .getSharedPreferences(Constants.PREF_FILE, Context.MODE_PRIVATE)

    private var appIndex: Int? = null
    private var currentApp: String? = null
    private var filter: String? = null
    private var lastApps: String? = null
    private var lastAppsSync: String? = null
    private var featureCount = 0

    @JvmStatic
    fun getFeatureCount(): Int {
        return prefs.getInt(Constants.EXTRA_FEATURE_USAGE, 0)
    }

    @JvmStatic
    fun setFeatureCount(featureCount: Int) {
        putInt(Constants.EXTRA_FEATURE_USAGE, featureCount)
    }

    @JvmStatic
    fun incrementFeatureCount(): Int {
        setFeatureCount(++featureCount)
        return featureCount
    }

    @JvmStatic
    fun getAppIndex(widgetId: Int): Int {
        return prefs.getInt(Constants.APP_INDEX + widgetId, 0)
    }

    @JvmStatic
    fun setAppIndex(appIndex: Int?, widgetId: Int) {
        if (this.appIndex == appIndex || appIndex == null) return
        putInt(Constants.APP_INDEX + widgetId, appIndex)
    }

    @JvmStatic
    fun setCurrentApp(currentApp: String?, widgetId: Int) {
        if (this.currentApp == currentApp) return
        putString(Constants.EXTRA_FEATURE_USAGE + widgetId, currentApp)
    }

    @JvmStatic
    fun getCurrentApp(widgetId: Int): String? {
        return prefs.getString(Constants.CURRENT_APP + widgetId, "")
    }

    @JvmStatic
    fun getFilter(widgetId: Int): String? {
        return prefs.getString(Constants.PREFS_FILTER_LIST + widgetId, "")
    }

    @JvmStatic
    fun setFilter(filter: String?, widgetId: Int) {
        if (this.filter == filter) return
        putString(Constants.PREFS_FILTER_LIST + widgetId, filter)
    }

    @JvmStatic
    fun getLastApps(widgetId: Int): String? {
        return prefs.getString(Constants.LAST_APPS + widgetId, "")
    }

    @JvmStatic
    fun setLastApps(lastApps: String?, widgetId: Int) {
        if (this.lastApps == lastApps) return
        putString(Constants.LAST_APPS + widgetId, lastApps)
    }

    @JvmStatic
    fun getLastAppsSync(widgetId: Int): String? {
        return prefs.getString(Constants.LAST_APPS_SYNC + widgetId, "")
    }

    @JvmStatic
    fun setLastAppsSync(lastAppsSync: String?, widgetId: Int) {
        if (this.lastAppsSync == lastAppsSync) return
        putString(Constants.LAST_APPS_SYNC + widgetId, lastAppsSync)
    }

    @JvmStatic
    fun getLastAppsSyncList(widgetId: Int): LinkedList<String> {
        val lastAppsSync = getLastAppsSync(widgetId)!!.split("\\|".toRegex())
        return LinkedList(lastAppsSync)
    }

    @JvmStatic
    fun setLastAppsSyncList(lastAppsSyncList: LinkedList<String>, widgetId: Int) {
        val lastAppsString = StringBuilder()
        for (app in lastAppsSyncList) {
            lastAppsString.append(app).append("|")
        }
        setLastAppsSync(lastAppsString.toString(), widgetId)
    }

    @JvmStatic
    fun addToLastAppsSync(newPackage: String, widgetId: Int) {
        val lastAppsList = getLastAppsSyncList(widgetId)
        if (lastAppsList.size > LAST_APPS_MAX_SIZE) lastAppsList.removeLast()
        lastAppsList.remove(newPackage)
        lastAppsList.addFirst(newPackage)
        setLastAppsSyncList(lastAppsList, widgetId)
    }

    @JvmStatic
    fun removeFromLastAppsSync(newPackage: String?, widgetId: Int) {
        val lastAppsList = getLastAppsSyncList(widgetId)
        lastAppsList.remove(newPackage)
        setLastAppsSyncList(lastAppsList, widgetId)
    }

    fun putString(key: String?, value: String?) {
        prefs.edit().putString(key, value).apply()
    }

    fun getString(key: String, value: String): String {
        return prefs.getString(key, value) as String
    }

    fun putInt(key: String?, value: Int) {
        prefs.edit().putInt(key, value).apply()
    }

    fun bool(key: String?): Boolean {
        return prefs.getBoolean(key, false)
    }

    fun clear() {
        prefs.edit().clear().apply()
    }

    fun runningThisFirstTime(tag: String?): Boolean {
        return if (!prefs.getBoolean(tag, false)) {
            prefs.edit().putBoolean(tag, true).apply()
            true
        } else {
            false
        }
    }
}